third/nmh/man/rmm.man
.\"
.\" %nmhwarning%
.\" $Id: rmm.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH RMM %manext1% MH.6.8 [%nmhversion%]
.SH NAME
rmm \- remove messages
.SH SYNOPSIS
.in +.5i
.ti -.5i
rmm
\%[+folder] \%[msgs]
\%[\-unlink] \%[\-nounlink]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
By default, \fIrmm\fR will remove the specified messages by renaming
the message files with preceding commas.  Such files will then need to
be removed in some manner after a certain amount of time.  Many sites
arrange for \fIcron\fR\0(8) to remove these files once a day, so check
with your system administrator.

Alternately, if you wish for \fIrmm\fR to really remove the files
representing these messages, you can use the `-unlink' switch.  But
messages removed by this method cannot be later recovered.

If you prefer a more sophisticated method of `removing' messages, you
can define the \fIrmmproc\fR profile component.  For example, you can
add a profile component such as

	rmmproc:	/home/coleman/bin/rmm_msgs

then instead of simply renaming the message file, \fIrmm\fR will call
the named program or script to handle the files that represent the
messages to be deleted.

Some users of csh prefer the following:

	alias rmm 'refile +d'

where folder +d is a folder for deleted messages, and

	alias mexp 'rm `mhpath +d all`'

is used to \*(lqexpunge\*(rq deleted messages.

The current message is not changed by \fIrmm\fR, so a \fInext\fR  will
advance to the next message in the folder as expected.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Current\-Folder:~^To find the default current folder
.Ps
^rmmproc:~^Program to delete the message
.Sa
refile(1), rmf(1)
.De
`+folder' defaults to the current folder
.Ds
`msgs' defaults to cur
.Ds
`-nounlink'
.Co
If a folder is given, it will become the current folder.
.Bu
Since \fIrefile\fR uses your \fIrmmproc\fP to delete the message,
the \fIrmmproc\fP must \fBNOT\fP call \fIrefile\fP without specifying
`\-normmproc', or you will create an infinte loop.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/rmm.c

/*
 * rmm.c -- remove a message(s)
 *
 * $Id: rmm.c,v 1.1.1.1 1999-02-07 18:14:16 danw Exp $
 */

#include <h/mh.h>

/*
 * We allocate space for message names and ranges
 * (msgs array) this number of elements at a time.
 */
#define MAXMSGS  256

static struct swit switches[] = {
#define UNLINKSW      0
    { "unlink", 0 },
#define NUNLINKSW    1
    { "nounlink", 0 },
#define VERSIONSW     2
    { "version", 0 },
#define	HELPSW        3
    { "help", 4 },
    { NULL, 0 }
};


int
main (int argc, char **argv)
{
    int nummsgs, maxmsgs, msgnum, unlink_msgs = 0;
    char *cp, *maildir, *folder = NULL;
    char buf[BUFSIZ], **argp;
    char **arguments, **msgs;
    struct msgs *mp;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Allocate the initial space to record message
     * names and ranges.
     */
    nummsgs = 0;
    maxmsgs = MAXMSGS;
    if (!(msgs = (char **) malloc ((size_t) (maxmsgs * sizeof(*msgs)))))
	adios (NULL, "unable to allocate storage");

    /* parse arguments */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown\n", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			  invo_name);
		print_help (buf, switches, 1);
		done (1);
	    case VERSIONSW:
		print_version(invo_name);
		done (1);

	    case UNLINKSW:
		unlink_msgs++;
		continue;
	    case NUNLINKSW:
		unlink_msgs = 0;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    /*
	     * Check if we need to allocate more space
	     * for message names/ranges.
	     */
	    if (nummsgs >= maxmsgs){
		maxmsgs += MAXMSGS;
		if (!(msgs = (char **) realloc (msgs,
			     (size_t) (maxmsgs * sizeof(*msgs)))))
		    adios (NULL, "unable to reallocate msgs storage");
	    }
	    msgs[nummsgs++] = cp;
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (!nummsgs)
	msgs[nummsgs++] = "cur";
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < nummsgs; msgnum++)
	if (!m_convert (mp, msgs[msgnum]))
	    done (1);
    seq_setprev (mp);		/* set the previous-sequence      */

    /*
     * This is hackish.  If we are using a external rmmproc,
     * then we need to update the current folder in the
     * context so the external rmmproc will remove files
     * from the correct directory.  This should be moved to
     * folder_delmsgs().
     */
    if (rmmproc) {
	context_replace (pfolder, folder);
	context_save ();
	fflush (stdout);
    }

    /* "remove" the SELECTED messages */
    folder_delmsgs (mp, unlink_msgs);

    seq_save (mp);		/* synchronize message sequences  */
    context_replace (pfolder, folder);	/* update current folder   */
    context_save ();			/* save the context file   */
    folder_free (mp);			/* free folder structure   */
    done (0);
}
